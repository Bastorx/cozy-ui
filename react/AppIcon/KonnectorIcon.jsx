import React from 'react'
import PropTypes from 'prop-types'

import { withClient } from 'cozy-client'

import { AppDoctype } from '../proptypes'
import AppIcon from './AppIcon'

/**
 * Used because AppIcon does not work well on mobile.
 * Extracted from Banks.
 *
 * TODO find out why and merge the two
 */
class KonnectorIcon extends React.PureComponent {
  constructor(props) {
    super(props)
    this.fetchIcon = this.fetchIcon.bind(this)
  }

  fetchIcon() {
    const { client, slug, app } = this.props
    return client.stackClient.getIconURL({
      type: 'konnector',
      slug: slug || app.slug
    })
  }

  render() {
    const { className } = this.props

    return <AppIcon fetchIcon={this.fetchIcon} className={className} />
  }
}

KonnectorIcon.propTypes = {
  client: PropTypes.object.isRequired,
  slug: PropTypes.string.isRequired,
  app: AppDoctype
}

export default withClient(KonnectorIcon)
