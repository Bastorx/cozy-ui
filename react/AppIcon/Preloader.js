const _loaded = {}

export const preload = async (app, domain, secure) => {
  if (!domain) throw new Error('Cannot fetch icon: missing domain')

  const source = _getAppIconURL(app, domain, secure)

  if (!source) {
    throw new Error(`Cannot get icon source for app ${app.name}`)
  }

  return new Promise((resolve, reject) => {
    const loader = document.createElement('img')
    loader.onload = () => {
      _loaded[source] = true
      resolve(source)
    }
    loader.onerror = () => {
      const loadError = new Error(`Error while preloading ${source}`)
      console.error(loadError.message)
      reject(loadError)
    }
    loader.src = source
  })
}

export const getPreloaded = (app, domain, secure) => {
  const source = _getAppIconURL(app, domain, secure)
  return source && _loaded[source] ? source : null
}

const _getProtocol = (secure = true) => {
  return `http${secure ? 's' : ''}:`
}

const _getAppIconURL = (app, domain, secure) => {
  if (!domain) return null

  let path = _getInstalledIconPath(app)
  path = path || _getRegistryIconPath(app)
  return path ? `${_getProtocol(secure)}//${domain}${path}` : null
}

const _getInstalledIconPath = app => {
  return app.links && app.links.icon
}

const _getRegistryIconPath = app => {
  return (
    app.latest_version &&
    app.latest_version.version &&
    `/registry/${app.slug}/${app.latest_version.version}/icon`
  )
}

export const Preloader = {
  preload
}

export default Preloader
