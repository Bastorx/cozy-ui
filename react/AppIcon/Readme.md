`<AppIcon />` renders an app icon.

It contains business logic related to apps documents.

Apps documents provided by the stack are containing a `links.icon` property which is public and can be directly used as `src` value for an `<img />` tag.

`<AppIcon />` is also able to retrieve an icon for a registry app. If no attribute `links.icon` is present on the app document, `<AppIcon />` tries to fetch the icon with a registry link, computed thanks to the app attributes `slug` and `latest_version.version`.

The given app can be:

* An object representing an app fetched from the registry
* An app's slug string

### Example

```jsx static
  <div>
    <AppIcon app={app} />
  </div>
```

## Using fetchIcon
Is it also possible to provide a custom asynchronous `fetchIcon` which takes an app document as argument and resolves with an icon URL (see [`URL.createObjectURL`](https://developer.mozilla.org/en/docs/Web/API/URL/createObjectURL)). `fetchIcon` may also throw errors.

### Example with fetchIcon

```jsx static
  <div>
    <AppIcon app={app} fetchIcon={icon => client.fetch(...)} />
  </div>
```
  </div>
```

## `domain` and `secure` props
If the `fetchIcon` is missing, the `<AppIcon />` component needs the `domain` prop (litteraly the cozy domain) and the secure prop, to know which protocol use between `http` or `https`.
