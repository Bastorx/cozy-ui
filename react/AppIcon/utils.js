const getProtocol = (secure = true) => {
  return `http${secure ? 's' : ''}:`
}

const getInstalledIconPath = app => {
  return app.links && app.links.icon
}

const getRegistryIconPath = app => {
  if (typeof app === 'string') {
    return `/registry/${app}/icon`
  }

  return (
    app.latest_version &&
    app.latest_version.version &&
    `/registry/${app.slug}/${app.latest_version.version}/icon`
  )
}

export const getAppIconURL = (app, domain, secure) => {
  let source
  if (!domain) throw new Error('Cannot fetch icon: missing domain')

  let path = getInstalledIconPath(app)
  path = path || getRegistryIconPath(app)
  source = path ? `${getProtocol(secure)}//${domain}${path}` : null

  if (!source) {
    throw new Error(`Cannot get icon source for app ${app.name}`)
  }
  return source
}
