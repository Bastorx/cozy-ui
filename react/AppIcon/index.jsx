import React, { Component } from 'react'
import PropTypes from 'prop-types'
import cx from 'classnames'
import styles from './styles.styl'

import Icon from '../Icon'

import appDefaultIcon from '../../assets/icons/ui/cube.svg'
import { getPreloaded, preload } from './Preloader'

const DONE = 'done'
const ERRORED = 'errored'
const FETCHING = 'fetching'

export class AppIcon extends Component {
  constructor(props, context) {
    super(props, context)
    const { app, domain, secure } = props
    const preloaded = getPreloaded(app, domain, secure)
    this.state = {
      error: null,
      icon: preloaded,
      status: preloaded ? DONE : FETCHING
    }
  }

  componentDidMount() {
    this.load()
  }

  async load() {
    const { app, domain, fetchIcon, onReady, secure } = this.props
    const loadFn = fetchIcon || preload
    let loadedUrl
    let loadError
    try {
      loadedUrl = await loadFn(app, domain, secure)
    } catch (error) {
      loadError = error
    }

    this.setState({
      error: loadError,
      icon: loadedUrl,
      status: loadError ? ERRORED : DONE
    })

    if (typeof onReady === 'function') {
      onReady()
    }
  }

  render() {
    const { alt, className } = this.props
    const { icon, status } = this.state
    switch (status) {
      case FETCHING:
        return (
          <div
            className={cx(
              styles['c-loading-placeholder'],
              styles['c-app-icon'],
              className
            )}
          />
        )
      case DONE:
        return (
          <img
            alt={alt}
            className={cx(styles['c-app-icon'], className)}
            src={icon}
          />
        )
      case ERRORED:
      default:
        return (
          <Icon
            className={cx(
              styles['c-app-icon'],
              styles['c-app-icon-default'],
              className
            )}
            height="100%"
            icon={appDefaultIcon}
            width="100%"
          />
        )
    }
  }
}

AppIcon.propTypes = {
  alt: PropTypes.string,
  app: PropTypes.object,
  className: PropTypes.string,
  domain: PropTypes.string,
  fetchIcon: PropTypes.func,
  onReady: PropTypes.func,
  secure: PropTypes.bool
}

export default AppIcon
