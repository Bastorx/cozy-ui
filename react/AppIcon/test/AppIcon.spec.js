'use strict'
/* eslint-env jest */

import React from 'react'
import { shallow } from 'enzyme'

import AppIcon from '../'

describe('AppIcon component', () => {
  const app = {
    links: {
      icon: 'apps/test/icon'
    }
  }

  const successFetchIcon = jest.fn().mockResolvedValue('data://test-url')

  it(`renders default when fetch error occurs`, () => {
    const failureFetchIcon = jest.fn().mockImplementation(() => {
      throw new Error('Mocked error')
    })
    const component = shallow(
      <AppIcon app={app} fetchIcon={failureFetchIcon} />
    ).getElement()
    expect(component).toMatchSnapshot()
  })

  it(`renders as loading`, () => {
    const component = shallow(
      <AppIcon app={app} fetchIcon={successFetchIcon} />
    ).getElement()
    expect(component).toMatchSnapshot()
  })

  it('renders app icon after fetching', done => {
    const wrapper = shallow(<AppIcon app={app} fetchIcon={successFetchIcon} />)

    setTimeout(() => {
      wrapper.update()
      const component = wrapper.getElement()
      expect(component).toMatchSnapshot()
      done()
    }, 10)
  })
})
