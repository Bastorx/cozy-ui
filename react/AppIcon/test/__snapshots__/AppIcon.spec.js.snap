// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppIcon component renders app icon after fetching 1`] = `
<img
  alt={undefined}
  className="styles__c-app-icon___2_O40"
  src="data://test-url"
/>
`;

exports[`AppIcon component renders as loading 1`] = `
<div
  className="styles__c-loading-placeholder___3L6Gz styles__c-app-icon___2_O40"
/>
`;

exports[`AppIcon component renders default when fetch error occurs 1`] = `
<Icon
  className="styles__c-app-icon___2_O40 styles__c-app-icon-default___3CEmt"
  height="100%"
  icon="test-file-stub"
  spin={false}
  width="100%"
/>
`;
